dbt_style_guide.md

```diff
@@ -1,237 +1,415 @@
 
-),
+- CTEs and SQL within a model should follow this structure:
+  - `with` statement
+  - Import CTEs
+  - Logical CTEs
+  - Final CTE
+  - Simple `select` statement
 
--- CTE comments go here
-filtered_events as (
+- All `{{ ref('...') }}` statements should be placed in CTEs at the top of the file ("import" CTEs)
 
-    ...
+- SQL should end with a simple select statement. All other logic should be contained within CTEs.  
+  Example: `select * from final`
 
-)
+- Where applicable, opt for filtering within import CTEs over filtering within logical CTEs.
 
-select * from filtered_events
-```
+- ### Example CTE
 
-## SQL style guide
+  ``` sql
+  with
 
-- Use trailing commas
-- Indents should be four spaces (except for predicates, which should line up with the `where` keyword)
-- Lines of SQL should be no longer than 80 characters
-- Field names and function names should all be lowercase
-- The `as` keyword should be used when aliasing a field or table
-- Fields should be stated before aggregates / window functions
-- Aggregations should be executed as early as possible before joining to another table.
-- Ordering and grouping by a number (eg. group by 1, 2) is preferred over listing the column names (see [this rant](https://blog.getdbt.com/write-better-sql-a-defense-of-group-by-1/) for why). Note that if you are grouping by more than a few columns, it may be worth revisiting your model design.
-- Prefer `union all` to `union` [*](http://docs.aws.amazon.com/redshift/latest/dg/c_example_unionall_query.html)
-- Avoid table aliases in join conditions (especially initialisms) – it's harder to understand what the table called "c" is compared to "customers".
-- If joining two or more tables, _always_ prefix your column names with the table alias. If only selecting from one table, prefixes are not needed.
-- Be explicit about your join (i.e. write `inner join` instead of `join`). `left joins` are normally the most useful, `right joins` often indicate that you should change which table you select `from` and which one you `join` to.
+  events as (
 
-- *DO NOT OPTIMIZE FOR A SMALLER NUMBER OF LINES OF CODE. NEWLINES ARE CHEAP, BRAIN TIME IS EXPENSIVE*
+      ...
+      where not is_deleted
 
-### Example SQL
-```sql
-with
+  ),
 
-my_data as (
+  -- CTE comments go here
+  events_joined as (
 
-    select * from {{ ref('my_data') }}
+      ...
 
-),
+  )
 
-some_cte as (
+  select * from events_joined
+  ```
 
-    select * from {{ ref('some_cte') }}
+## SQL style guide
+- **DO NOT OPTIMIZE FOR A SMALLER NUMBER OF LINES OF CODE.**  
+  New lines are cheap, brain time is expensive; new lines should be used within  
+  reason to produce code that is easily read.
 
-),
+- Use trailing commas
 
-some_cte_agg as (
+- Indents should be four spaces 
 
-    select
-        id,
-        sum(field_4) as total_field_4,
-        max(field_5) as max_field_5
+- When dealing with long `when` or `where` clauses, predicates should be on a new

```

Comment:
Is this not already covered by the section below:

> - When filtering by multiple clauses, each clause and expression should be on a new line.  

Might be good to connect these two?

Reply:
Removed the last one! [Commit](https://github.com/dbt-labs/corp/pull/63/commits/d743e54439bf4b269d94927a6c48bb937e9eb338)